### PR TITLE
[RunAllTests] Fix part of #5343: Upload generated code coverage report as comments

### PR DESCRIPTION
**Pull request by Rd4dev**
## Explanation

Fixes part of #5343 

### Project
[PR 2.4 of Project 4.1]

### Changes Made

With the protos collected and stored with their respective file paths through #5465, this PR will focus on collecting the stored proto files and passing it to the script to generate a coverage report along with a status check to decide the success failure case of the CI Coverage Check run.

**Collection of protos**
- The stored protos are now being uploaded as an artifact.
- The name of the artifact is set to a dynamic value corresponding to its shard_name.
  - The shard_name is extracted from the first portion of the `CHANGED_FILES_BUCKET_BASE64_ENCODED_SHARD` value.
    - eg: The matrix job - [app-shard0;H4sIAAAAAAAAAONiTiwoEHIFEvrFRcn6uYmZefpZi....] will save the artifact with name **coverage-report-app-shard0**
    - The dynamic or unique name is important to prevent the following error: 
```
Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: 
an artifact with this name already exists on the workflow run
```
- The artifacts that match the pattern of coverage-report-* are downloaded.

```
Found 4 artifact(s)
Filtering artifacts by pattern 'coverage-report-*'
Preparing to download the following artifacts:
- coverage-report-domain-shard0 (ID: 1799350592, Size: 489)
- coverage-report-scripts-shard2 (ID: 1799348996, Size: 2075)
- coverage-report-app-shard3 (ID: 1799348891, Size: 770)
- coverage-report-generic-shard1 (ID: 1799348719, Size: 551)
```
- All the stored coverage_report.pb files are found and their paths are stored as list.
- The list of coverage_report.pb files are then passed to the CoverageReporter.kt script to handle the coverage report protos.
- The script combines them as one single CoverageReportContainer to generate a Markdown report with it.
- Finally, a .md report is stored and based on the status of the coverage check, the job either succeeds or fails.
- The stored md report is again uploaded and downloaded as an artifact to pass it for publication of the report.
- The CoverageReport.md file is then uploaded as a comment to the corresponding PR using - https://github.com/peter-evans/create-or-update-comment action

>The comment will be published regardless of whether the coverage passes or fails. The only time comments will not be published is if the unit tests themselves fail, making the coverage checks non-functional (also less cluttered). 

#

### CI Run Data

This check was run with 4 different shards on all possible cases capable of being produced.
Check : https://github.com/oppia/oppia-android/actions/runs/10340867060?pr=5469
Comment Published with the above run: https://github.com/oppia/oppia-android/pull/5469#issuecomment-2282804068

The above cases are run with custom temporary changes to the test exemption file (testing purposes).
```
test_file_exemption {
  exempted_file_path: "scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt"
  override_min_coverage_percent_required: 101
}
test_file_exemption {
  exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/ConsoleLogger.kt"
  override_min_coverage_percent_required: 30
}
```

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by Rd4dev** [comment]
@BenHenning, the error on `overload resolution ambiguity:` is yet to be addressed, but can you PTAL on the implementations for uploading the comments? I have also added tests for the script modifications.

File: .github/workflows/code_coverage.yml
Comment (BenHenning) [inline .github/workflows/code_coverage.yml]:
```suggestion
      - name: Check that coverage status is passed
```

File: scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt]:
```suggestion
/* ANSI escape codes for colors. */

```

This isn't a KDoc (since it has no element to document.

File: scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt:32]: Nice catch.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @Rd4dev! Just had a few comments for the base PR, but also some follow-up items:
1. It seems that the ``main()`` ambiguity issue is coming from adding the new ``main()`` function to ``CoverageReporter`` since ``RunCoverage`` depends on it here: https://github.com/oppia/oppia-android/blob/f1e3463f0ab4f39df20220efea7a2474fa663556/scripts/src/java/org/oppia/android/scripts/coverage/BUILD.bazel#L16
2. Please make sure to update this PR to account for the additional requests from the product review meeting yesterday (so that we can consider this project functionally complete after this PR is merged).
3. Once the ``main()`` is resolved, please update the PR title to run with ``[RunAllTests]`` so that we can verify coverage for all tests in the codebase. Now that you've verified coverage working for some targets (per the PR's comment history), we ought to keep 'RunAllTests' on for the remainder of the PR's life since we'll want to continually try all tests to make sure develop doesn't break once this PR is merged.
4. As part of (3) you'll probably need to add some additional test exemptions for tests that pass but have below 70% coverage (I'm positive there are at least a few from my own analysis). I suggest doing the following: finish the rest of the PR (including fixing the main issue) and waiting until CI finishes with a pass, then add 'RunAllTests' and re-run CI by closing and reopening the PR, once coverage runs without additional fixes being needed then update the PR to include any additional exemptions that are needed and verify CI fully passes.

File: scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt
Comment (Rd4dev) [reply scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt]: Done, updated as suggested.

File: .github/workflows/code_coverage.yml
Comment (Rd4dev) [reply .github/workflows/code_coverage.yml]: Done, updated as suggested.

File: scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt
Comment (Rd4dev) [reply scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt:32]: Thanks :)

**Comment by Rd4dev** [comment]
@BenHenning, also just found another issue with the existing implementation, since the CoverageReporter was moved to a new package, it was not computed properly, that needs to be fixed. Will do that once RunAllTests complete. 

Edit:
One thing I suspect, would performing any merge with other branches affect this? - Not really sure why.
But while fixing ambiguity error, I pushed a fix, then reverted to a previous commit, then switched to a new branch and merged changes from there, would these cause any issues, is this even allowed? 

I think that messed the commit history earlier and I doubt if I can proceed with this PR, after RunAllTests can I just create a new PR so this doesn't affect anything else.

**Comment by Rd4dev** [comment]
@BenHenning, the PR was tested with [RunAllTests] and the initial report was generated:  https://github.com/oppia/oppia-android/pull/5469#issuecomment-2286075490, with this the test exemptions to introduce overriding coverage percentages for all the files that failed below the min threshold are added to the test_file_exemptions.textproto file and ran again with [RunAllTests] to successfully Pass all coverage checks: https://github.com/oppia/oppia-android/pull/5469#issuecomment-2286506104.

**Additional fixes:** 
- The previously mentioned `main` ambiguity error was resolved by restructuring the file that had another main function in the same package to its sub package.

- CoverageReporter rename / move issue mentioned was resolved by taking into account all the renamed files. And their related test files are also added to the `GitClientTest.kt`. But I would like to check if this test case - `testUtility_featureBranch_movedFile_stagedAndCommitted_returnsNewFile` in the `computeChangedFilesTest` was functioning right? with the added functionality `moveFile`, since it did pass even without the new change to include the move files in the computed files. (But it failed with staging the oldFilePath, but with the new changes it accepts staging oldFilePath). Would like to confirm if the test case is functioning right.

- Argument list too long issue - was resolved by moving away from saving the pb_file list in ENVs to their own file and passing the file.txt to the script, which is then parsed in the script. (Added test cases for the same).

- Updated the template based on the M2 demo.

Can you PTAL?

**Review by BenHenning — APPROVED** [review]
Thanks @Rd4dev! I had some follow-up nits, but decided to drop them since they are minor and I'd rather see this get merged.

It is _fantastic_ to see the CI run working, and I'm amused to see how many shards 'RunAllTests' produces now.

The main thing left is to update the template to link to the new wiki article, but perhaps that is better to do in the wiki PR itself since it won't exist until then (plus we can them prioritize getting this merged).

**Comment by BenHenning** [comment]
Holding off merge until CI is green (can't auto-merge since the new workflows won't be marked as blocking until this PR is merged).

**Comment by BenHenning** [comment]
It actually occurs to me that this _could_ be the perfect moment to verify the new final "check code coverage results" workflow passes when expected, so I decided to mark it as blocking and enable auto-merge. If it works, the PR will be merged. :)

**Pull request merged**
